src/hotspot/cpu/riscv/globals_riscv.hpp

```diff
@@ -82,19 +82,30 @@ define_pd_global(intx, InlineSmallCode,          1000);
                                                                                  \
   product(bool, NearCpool, true,                                                 \
          "constant pool is close to instructions")                               \
+  product(bool, UseBlockZeroing, false,                                          \
+          "Use Zicboz for block zeroing")                                        \
   product(intx, BlockZeroingLowLimit, 256,                                       \
           "Minimum size in bytes when block zeroing will be used")               \
           range(1, max_jint)                                                     \
+  product(intx, CacheLineSize, DEFAULT_CACHE_LINE_SIZE,                          \
```

Comment:
Is there is no guarantee that a platform will provide both `Zicboz` _and_ `Zic64b` even if the cache lines are 64 bytes? As in, the cache lines are 64 bytes, but the CPU doesn't notify the kernel that it supports `Zic64b`?

In that case, we won't use `Zicboz` which can be significantly faster than other approaches to zero out memory (the CPU is free not to fetch a cache line before storing it, since the whole cache line is zeroed out). Given the overall cost of zeroing out memory in Java, it's necessary to ensure it's as fast as possible.

I overall agree with the 64 bytes is an industry standard. What I want to ensure is that we can eagerly enable `UseBlockZeroing` whenever we can.

Reply:
Another solution is to do something similar to AArch64 and have vendor-specific setup in vm_version. See https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/vm_version_aarch64.cpp#L132-L222 for example. `CacheLineSize` can then be then be a global variable defined in `src/hotspot/cpu/riscv/globalDefinitions_riscv.hpp` which is set based on the vendor and a default value.

Reply:
> Is there is no guarantee that a platform will provide both `Zicboz` _and_ `Zic64b` even if the cache lines are 64 bytes? As in, the cache lines are 64 bytes, but the CPU doesn't notify the kernel that it supports `Zic64b`?
> 
> In that case, we won't use `Zicboz` which can be significantly faster than other approaches to zero out memory (the CPU is free not to fetch a cache line before storing it, since the whole cache line is zeroed out). Given the overall cost of zeroing out memory in Java, it's necessary to ensure it's as fast as possible.
> 
> I overall agree with the 64 bytes is an industry standard. What I want to ensure is that we can eagerly enable `UseBlockZeroing` whenever we can.

AFAIK, it's still not clear how those extensions like Zva, Zvb, Zicboz, etc, should be auto detected at runtime for now. 
I see this is still under discussion on LKML not long ago. So for now we depend on users to specify corresponding JVM options explicitly on the command line when available. So for your concern here, I would expect the user to specify both Zicboz and Zic64b explictly for performance. We will revisit the auto-detection part once the correct way is settled down.

Reply:
We can already do auto-detection via `/proc/cpuinfo`. I agree it's not the best way nor the most stable way, and I can only assume there will be a better way, but there is already a way.

I'll remove this `CacheLineSize` flag for now and we'll revisit later.

Reply:
> We can already do auto-detection via `/proc/cpuinfo`. I agree it's not the best way nor the most stable way, and I can only assume there will be a better way, but there is already a way.

I think I mean extensions like 'Zba', 'Zbb', etc here.
Are you sure those could be detected from `/proc/cpuinfo`?
I am assuming that only works for extensions contained in RV64GCV (IMAFDCV).

> I'll remove this `CacheLineSize` flag for now and we'll revisit later.

Reply:
This is a sample of `/proc/cpuinfo` on a QEMU:
```
processor	: 0
hart		: 30
isa		: rv64imafdcvh_zicsr_zifencei_zihintpause_zba_zbb_zbc_zbs_sstc
mmu		: sv48
```

You can `strstr` (or some alternative of it) on the `isa` string for the extension you want. That `isa` string comes from the kernel at https://github.com/torvalds/linux/blob/master/arch/riscv/kernel/cpu.c#L141-L146. The obvious problem at the moment is that not all extensions have been added here which is why we will want a better detection mechanism. However, on custom kernels, you can have most of the extensions you want.

Reply:
@RealFYang I've done the change as you requested.

Reply:
Thanks for the update. Looks like DEFAULT_CACHE_LINE_SIZE is not defined anywhere? Looks good, otherwise.

Reply:
It's defined in `src/hotspot/share/utilities/globalDefinitions.hpp` but let me add it in `src/hotspot/cpu/riscv/globalDefinitions_riscv.hpp` just to make sure.

Reply:
> I overall agree with the 64 bytes is an industry standard. What I want to ensure is that we can eagerly enable `UseBlockZeroing` whenever we can.

I strongly disagree.

Until processor reports Zic64b supported we should assume the cache line size could be anything (2^N). I'm running some tests on risc-v fpga core and it has 16-bytes cache block size.
having to patch ( change default cache line size) and rebuild openjdk to just try new Zicbo{z,p,m} opcodes ( once they are implemented) - isn't a nice way

Reply:
> > I overall agree with the 64 bytes is an industry standard. What I want to ensure is that we can eagerly enable `UseBlockZeroing` whenever we can.
> 
> I strongly disagree.
> 
> Until processor reports Zic64b supported we should assume the cache line size could be anything (2^N). I'm running some tests on risc-v fpga core and it has 16-bytes cache block size. having to patch ( change default cache line size) and rebuild openjdk to just try new Zicbo{z,p,m} opcodes ( once they are implemented) - isn't a nice way

Now I see the concern here. I agree to have the an option for cache-line size in that respect.

Reply:
I'll revert to what I had to set CacheLineSize then.